Implement SQLCommenter support

iand675 opened the pull request:
This adds support for the [`sqlcommenter`](https://google.github.io/sqlcommenter/) spec, as outlined by folks at Google. We're using it at https://github.com/MercuryTechnologies for improved correlation between SQL queries and the rest of the OTel ecosystem. The main area we're seeing benefits is via https://pganalyze.com/, which has native support for the format. Namely, we can have "explain analyze" traces added to our existing application traces to help us understand why certain queries behave badly in production.

We've used this in prod for about 10 months, so I'm not particularly concerned about the efficacy of the implementation, but I do want to ensure that it builds against API changes & is updated on a similar cadence to other packages in this repo.

Why should this package live in this repo? Because Google donated it to the OpenTelemetry community: https://cloud.google.com/blog/products/databases/sqlcommenter-merges-with-opentelemetry

iand675 committed:
Update api version

iand675 committed:
Rerelease

Commit:
cache

iand675 committed:
Merge pull request #15 from kakkun61/cache

iand675 committed:
GHC 9.2 support

Commit:
Nix for CI

Commit:
Merge branch 'main' into nix

iand675 committed:
Merge pull request #18 from kakkun61/nix

iand675 committed:
add cloudflare to repo

lf- committed:
Fix copypasted docs for alwaysOff (#19)

claymager committed:
copy into hs-opentelemetry

claymager committed:
Prettification

- Deletion of unused imports and unrelated commented-out code
- Minor refactor of inSpanM''
- run code formatter

claymager committed:
Remove more references to cloudflare

claymager committed:
Add reference to potential issues in README

claymager committed:
Include updates to generated cabal file

claymager committed:
reset tooling files

psilospore committed:
Updated docs to use makeTracer instead of deprecated getTracer

psilospore committed:
Merge pull request #1 from psilospore/psilospore-patch-1

Updated docs to use makeTracer instead of deprecated getTracer

lf- committed:
bugfix: Batch processor did not shut down cleanly, losing metrics

When the process exited, the sending of metrics to OTLP was rudely
interrupted by `cancel`. This meant that the root span would often go
missing if you instrumented some kind of command line application with
the root span created at process creation.

This patch changes the shutdown logic to first ask the exporter to send
what it has and shut down, then forcibly kill it if that takes too long.

lf- committed:
Add hspec instrumentation

It would be useful to have instrumentation of test suites. My
original use case for this is to validate a change in Persistent
instrumentation, for which hitting it really hard with the Mercury test
suite seems like a reasonable strategy.

lf- committed:
Add Show instances

I needed these for debugging, but they are generally useful.

There's an unfortunate amount of trailing whitespace deletion in this
diff. Sorry :(

parsonsmatt committed:
Support GHC 9.4

iand675 committed:
Merge pull request #25 from MercuryTechnologies/show-instances

iand675 committed:
Merge pull request #27 from parsonsmatt/matt/support-ghc-94

iand675 committed:
Merge pull request #22 from psilospore/main

iand675 committed:
Merge pull request #23 from MercuryTechnologies/batch-processor-shutdown

iand675 committed:
Merge pull request #21 from BlueChipFinancial/utils-exceptions

iand675 committed:
version bumps

lf- committed:
Add an example and docs

iand675 committed:
Merge pull request #24 from MercuryTechnologies/hspec-instrumentation

iand675 committed:
Add Jade as signficant contributor

lf- committed:
Fix accidental italics in docs

lf- committed:
Merge pull request #30 from MercuryTechnologies/fix-italics

Fix accidental italics in docs

kakkun61 committed:
format action

kakkun61 committed:
add fourmolu

kakkun61 committed:
fourmolu without nix

Nixpkgs release-22.05 channel's fourmolu is too old to parse some codes.

kakkun61 committed:
fourmolu parsable

kakkun61 committed:
format

kakkun61 committed:
proto-lens 0.7.1.2 already supports GHC 9.4

https://github.com/google/proto-lens/pull/440

kakkun61 committed:
fix stack compilation error

kakkun61 committed:
fix github action

kakkun61 committed:
update cabal files

kakkun61 committed:
config nix path

iand675 committed:
Merge pull request #31 from herp-inc/fix-compilation-error

kakkun61 committed:
add c libs

kakkun61 committed:
CI for cabal

kakkun61 committed:
cabal under nix

kakkun61 committed:
update hie.yaml

kakkun61 committed:
hpack

kakkun61 committed:
CI runs also on PR

kakkun61 committed:
Merge pull request #34 from herp-inc/ci-pr

CI runs also on PR

kakkun61 committed:
Merge branch 'main' into fix-compilation-error-cabal

kakkun61 committed:
cabal update

kakkun61 committed:
nix-shell

kakkun61 committed:
Merge pull request #33 from herp-inc/fix-compilation-error-cabal

Fix compilation error for cabal users

kakkun61 committed:
phony

kakkun61 committed:
indent-wheres: true

kakkun61 committed:
format

iand675 committed:
Merge pull request #35 from herp-inc/format

lf- committed:
Fix typo in docs

lf- committed:
Merge pull request #36 from MercuryTechnologies/fix-typo

Fix typo in docs

Commit:
add instances for monad-tracer

Commit:
StateVar and Only need libffi

Commit:
libffi for stack

parsonsmatt committed:
Fix batch shutdown deadlock

parsonsmatt committed:
ahh

parsonsmatt committed:
space

parsonsmatt committed:
manually fix formatting

parsonsmatt committed:
wow

parsonsmatt committed:
Synchronously cancel the worker

parsonsmatt committed:
Flush queue on shutdown

parsonsmatt committed:
error return refactor

iand675 committed:
Merge pull request #42 from parsonsmatt/mattp/fix-batch-shutdown

Fix batch shutdown deadlock

iand675 committed:
Release SDK and OTLP exporter shutdown fixes

lf- committed:
Add a vendor-honeycomb package with trace link stuff (#39)

* wip vendor honeycomb support

* Add a little testing

* Remove monad environment stuff

* Fix timestamp calculation

* Expose a super simple API

* correct test

* New approach using the Context vault to deal with the target

* format

* format ii

lf- committed:
sdk: Crash if -threaded is unset (#46)

* sdk: Crash if -threaded is unset

registerDelay will throw an exception, which we will silently swallow, if we
don't preemptively just blow up the program. It's never correct to start up
without a threaded runtime.

* Add a note

Commit:
overlappable

Commit:
update CI

Commit:
format with fourmolu 0.10.1.0

Commit:
freeze

Commit:
init a Datadog propagator

Commit:
Merge pull request #48 from herp-inc/update-ci

Update CI

Commit:
init persistent-mysql

Commit:
Merge remote-tracking branch 'upstream/main' into datadog-propergator

Commit:
format

Commit:
hide constructors of IDs

https://github.com/iand675/hs-opentelemetry/pull/49#discussion_r1070138527

iand675 committed:
Merge pull request #37 from herp-inc/monad-tracer-instances

add instances for monad-tracer

Commit:
add a datadog propagator to README

iand675 committed:
Merge pull request #49 from herp-inc/datadog-propergator

Datadog propagator

Commit:
add persistent-mysql to README

Commit:
fix GHC version tag about 8.10

kakkun61 committed:
persistent-mysql attributes

kakkun61 committed:
Workaround

https://github.com/haskell/actions/issues/187

kakkun61 committed:
fix CI name

9999years committed:
Update instrumentation-hspec for GHC 9.6

9999years committed:
Import `join` from `Control.Monad`

Reexports removed in mtl-2.3/transformers-0.6

9999years committed:
Fix memory leak

Co-authored-by: Ian-Woo Kim @wavewave

iand675 committed:
Merge pull request #53 from herp-inc/workaround-ghcup

Workaround for ghcup issue

solomon-b committed:
Adds B3 Propagator Library

Commit:
Merge pull request #63 from 9999years/honeycomb

Fix `hs-opentelemetry-vendor-honeycomb` for mtl-2.3/transformers-0.6

solomon-b committed:
Apply Fourmolu Formatting

solomon-b committed:
Add b3 propagator to stack.yaml

Commit:
Merge pull request #62 from 9999years/instrumentation-hspec-ghc96

Update instrumentation-hspec for mtl-2.3/GHC-9.6

solomon-b committed:
Remove unused 'case-insensitive' depenednecy from b3 propagator

Commit:
Merge pull request #66 from hasura/feature/adds-b3-propagator

Adds B3 Propagator Library

kakkun61 committed:
more familiar and better perf. "show" instance

Commit:
refresh CI cache

Commit:
refresh CI cache

Commit:
show emptyAppendOnlyBoundedCollection

Commit:
Merge branch 'refresh-ci-cache' into better-show

Commit:
format

Commit:
fix compilation errors

Commit:
use bang patterns instead of seq

9999years committed:
Merge pull request #1 from herp-inc/api-fix-memory-leak

Fix compilation errors etc.

iand675 committed:
Merge pull request #55 from herp-inc/fix-ci-name

fix a CI step's name

iand675 committed:
Merge pull request #67 from herp-inc/better-show

more familiar and better performance "show" instance

iand675 committed:
Merge pull request #64 from 9999years/api-fix-memory-leak

Fix memory leak

iand675 committed:
Work on getting the nix setup usable for building & testing the codebase in CI

iand675 committed:
Add linkFarm default target

iand675 committed:
flake tweaks

iand675 committed:
test

iand675 committed:
fiddle

iand675 committed:
fiddle

iand675 committed:
maintenance

iand675 committed:
maintenance

iand675 committed:
regen cabal files

iand675 committed:
fix flake support for GHC 9.4

iand675 committed:
try flake in GHA

iand675 committed:
quotes

iand675 committed:
morale must improve

iand675 committed:
this should probably definitely work

iand675 committed:
fiddle

iand675 committed:
fiddle

iand675 committed:
this is promising

iand675 committed:
cabal

iand675 committed:
remove constraints from cabal.project

iand675 committed:
Debugging

iand675 committed:
Debugging

iand675 committed:
Don't attach empty contexts to WAI threads when finished

iand675 committed:
bump thread-utils-context

iand675 committed:
bip

iand675 committed:
not sure if post qualified works in GHC 8.x

iand675 committed:
Small documentation improvements

iand675 committed:
Cabal file wrangling

iand675 committed:
Version bumps

iand675 committed:
Add vendors/honeycomb to cabal project

iand675 committed:
Release SDK with b3 support

iand675 committed:
Fix compilation error

JonathanLorimer committed:
Fix README.md code example

iand675 committed:
Merge pull request #72 from JonathanLorimer/JonathanLorimer-patch-1

Fix README.md code example

iand675 committed:
Mercury work

iand675 committed:
Fix build?

iand675 committed:
fiddle

parsonsmatt committed:
Add instrumentSpec function

parsonsmatt committed:
format

iand675 committed:
Merge pull request #81 from parsonsmatt/mattp/hspec-describe-context

Add instrumentSpec function

iand675 committed:
Merge remote-tracking branch 'origin/main' into improved-dux

iand675 committed:
fold in hspec changes and fix the examples

chrismanning committed:
Add missing import

chrismanning committed:
Decode percent-encoded baggage headers

chrismanning committed:
Formatting

chrismanning committed:
Only import word64ToWord# for base-4.17.0 or newer

jkeuhlen committed:
Implement maxExportBatchSize in batch processor

Commit:
Merge pull request #84 from chrismanning/decode-percent-encoded-baggage

Decode percent-encoded baggage headers

chrismanning committed:
Formatting

chrismanning committed:
Formatting

Commit:
Merge pull request #83 from chrismanning/main

Add missing import

iand675 committed:
Merge pull request #85 from jkeuhlen/ImplmentMaxExportBatchSize

Implement max export batch size

Commit:
newOpenTelemetryWaiMiddleware' is now not IO action

Commit:
update change log

Commit:
Merge pull request #86 from iand675/wai-no-io

newOpenTelemetryWaiMiddleware' is now not an IO action

Commit:
Merge pull request #75 from herp-inc/stack-ghc-8.10

fix GHC version tag about 8.10

Commit:
make Attributes be HashMap

Commit:
fourmolu 0.13.1.0

Commit:
change log and version up

Commit:
Merge branch 'main' into attributes-hashmap

iand675 committed:
Merge pull request #88 from herp-inc/attributes-hashmap

Use maps instead of lists of pairs as attributes

Commit:
Merge pull request #87 from herp-inc/update-fourmolu

Use fourmolu 0.13.1.0

Commit:
Merge remote-tracking branch 'upstream/main' into persistent-mysql

Commit:
fix attributes type

Commit:
Merge pull request #50 from herp-inc/persistent-mysql

persistent-mysql instrumentation

iand675 committed:
back to stack

iand675 committed:
fiddle

iand675 committed:
fiddle

iand675 committed:
newer nix version

iand675 committed:
oodles of formatting

iand675 committed:
fourmolu action moved to haskell-actions

iand675 committed:
fiddle

iand675 committed:
Fix compilation on GHC 8.12

iand675 committed:
get unsafeCoerce# from GHC.Exts

iand675 committed:
format

iand675 committed:
Merge remote-tracking branch 'origin/main'

iand675 committed:
tag releases

parsonsmatt committed:
Swallow localhost errors

parsonsmatt committed:
lol

iand675 committed:
Merge pull request #89 from parsonsmatt/mattp/swallow-localhost-errors

Swallow localhost errors

Qqwy committed:
Fix git conflict in hs-opentelemetry-utils-exceptions.cabal

There was a committed git conflict in utils/exceptions/hs-opentelemetry-utils-exceptions.cabal

This commit fixes it.

iand675 committed:
Merge pull request #107 from Qqwy/Qqwy-patch-1

Fix git conflict in hs-opentelemetry-utils-exceptions.cabal

evanlauer1 committed:
Update HTTP Semantic Conventions to Stable Versions (#118)

* Work in Progress: Request and Response Attributes updated to match the latest stable semantics if the environment variable OTEL_SEMCONV_STABILITY_OPT_IN is "http"

* Added Stable attributes to instrumentResponse

* Refactored to fit new httpTracerProvider return type

* Fixed missing "do" notation and removed unnecessary imports

* Fix typo and update comment

* Found unchanged http name

* Found outdated http attribute names

* Fixed cradle so HLS works

* Added Settings File for application-wide environment variable configuration

* Moved semConvStabilityOptIn and related types to OpenTelemetry.Settings

* Changed code using HttpTracerProvider to work with changed return type

* Fixed exports

* Replaced general Settings.hs with specific SemConvStability.hs

* Removed comment because the problem it referenced is solved

* removed unused language pragma

* fixed imports

* Replaced outdated http conventions based on the value of OTEL_SEMCONV_STABILITY_OPT_IN

* changed line about generation version

* added unordered-containers to dependencies

* Moved semConvStabilityOptIn so that it is in scope

* Updated dependencies for testing

* WIP testing for http-client

* Generated by newer version

* Removed WIP testing

* Updated getSemConvStabilityOptIn  so it parses environment variable as a comma-separated list

* derives Show  and Eq for testing purposes

* Added tests for SemConvStabilityOptIn to make sure environment variable is read correctly

* Updated Haddock to include blurb about the new HTTP semantic conventions

* Removed hspec from dependencies

* Auto-generated change. Removed hspec from build-depends

* Added newline to end of file

* Split off parsing logic from getSemConvStabilityOptIn to parseSemConvStabilityOptIn

* Added blank line for readability

* Changed pure to Just to improve readability and contrast with Nothing

* Rewrote tests to eliminate boilerplate and increase readability

* Refactored and Renamed SemanticsConfig.hs to future proof

* Refactored SemanticsConfig so options are Enums for pattern matching instead of booleans

* Updated SemanticsConfig tests. NOTE: does not currently pass memoization tests.

* Updated to match changes to SemanticsConfig.hs

* Fix infinite loop issue

Co-authored-by: Ian Duncan <ian@iankduncan.com>

* changed from . to $

* Added haddocks

* Formatting fixes

---------

Co-authored-by: Ian Duncan <ian@iankduncan.com>

iand675 committed:
Add sqlcommenter package

Gabriella439 committed:
Export `FlushResult` (#96)

The return value of `forceFlushTracerProvider` is `FlushResult`, but
that type isn't exported so it can't be used, and this change fixes
that.

pbrisbin committed:
Add Datadog as a Known Propagator (#117)

* Don't depend on sdk from propagator-datadog

This meant adding ourselves to `knownPropagators` would be circular. We
can easily grab this type from the `api` package instead.

* Add datadog as a known propagator

iand675 committed:
CI overhaul (#120)

* test

* Add some more GHC versions

* Add pre-commit checks and some other goodies

* missed some hooks there

* Exclude shell builds for now

* GHC 8.10 and 9.0 aren't in nixpkgs anymore?

* Get particular about build packages

* Don't torture garnix with too many packages

* fix glob

* Hardcode packages since globs are wimpy

* pre-commit checks

* turn off old checks

* Get checks to agree with each other

iand675 committed:
Add basic CONTRIBUTING.md doc

evanlauer1 committed:
Change OTLP generation strategy from manual to auto-generation (#122)

* Experimenting with protoc autogen

* Got first pass at auto-generated files working

* Auto-generated files

* Moved proto submodule into otlp

* Fixed custom setup to work with Cabal using defaultMain

* added field "autogen-modules"

* changed verbatim generated-exposed-modules

* Formatting

* Added documentation for auto-generation

* Further updated auto-generation instructions

* tweak

* experimental CI fix

* update cabal files

---------

Co-authored-by: Ian Duncan <ian@iankduncan.com>

evanlauer1 committed:
Rename `Log` to `LogRecord` and update data type to match logs data model spec (#125)

* Created `LogAttributes` where `AnyValue` is a superset of `Attribute`

* Renamed `Log` to `LogRecord` and updated fields to match current Log Data Model spec

evanlauer1 committed:
Created emitLogRecord that takes a Logger as an argument. Refactored types to a separate module. stab-267 (#126)

* Created emitLogRecord that takes a Logger as an argument. Refactored types to a separate module

* Deleted legacy implementations of SeverityNumber

* Updated export list

evanlauer1 committed:
Added instrumentationScope to Logger and LogRecord (#127)

* Added instrumentationScope to Logger and LogRecord

* Removed Hashable instance from InstrumentationLibrary and Attributes

* Added Hashable instance back to InstrumentationLibrary

* Updated comment for instrumentationLibrary

* Further updated comment

evanlauer1 committed:
Add LoggerProvider to create Loggers and allow registering a global LoggerProvider (#128)

* add-loggerProvider-to-create-loggers

* Added documentation

* Removed `Maybe` from Resources field because `emptyResources` can be used instead

* Removed MonadIO constraint from `createLoggerProvider`

evanlauer1 committed:
Refactored attribute limits to be controlled by the LoggerProvider (#129)

michaelpj committed:
Fix callerAttributes in the presence of frozen call stacks (#137)

Freezing the call stack is common when writing your own utility wrappers
around e.g. `withSpan`. In this case you want to freeze the call stack
before calling your utility, so that `withSpan` and friends see the
"application" function as the top of the call stack.

However, the functions for getting information from call stacks
just generally didn't work right in the presence of
frozen call stacks. They assumed that the call stack was
unfrozen, and hence that they could look up a fixed number of steps to
find the caller info.

We can't in general tell directly if the call stack is frozen, but we
use a pretty decent heuristic here: is the top element exactly our own
function? If so, we probably have an unfrozen call stack. Otherwise we
might have a frozen call stack and we have to try and do something else.

iand675 committed:
Otel overlay (#138)

* testing

* tweak

evanlauer1 committed:
Added Logging when creating spans or emitting logs drops attributes. Updated Severity Number (#133)

* Added Logging to emitLogRecord and createSpanWithoutCallStack when dropping attributes. Updated Severity Number

* Added comments warning users not to use emitLogRecord

* Added further documentation to warn against using emitOTelLogRecord

* Explicitly handle Unknown case in toShortName

* Changed Err to Error in SeverityNumber

* Moved Logging.Core to Internal.Logging.Core to hide functions

* Revert "Moved Logging.Core to Internal.Logging.Core to hide functions"

This reverts commit 32c5e58978aef92d94bc9042a471340daafd6002.

* Moved Logging.Core to Internal.Logging.Core

* Re-added Logging.Core with re-exported functions from Internal.Logging.Core

evanlauer1 committed:
Refactored LogRecords to be mutable so that attributes can be modified after creation (#131)

* Refactored LogRecords to be mutable so that attributes can be modified after creation

* Updated test suite

* Added ReadableLogRecord and ReadWriteLogRecord interfaces

* Added typeclasses for ReadableLogRecords and ReadWriteLogRecords

* Replaced type variable in LogRecord with AnyValue

* Changed Readable and ReadWriteLogRecord to be concrete types

* Changed mkReadableLogRecord to take in a ReadWriteLogRecord

* Added NullValue to AnyValue and removed body parameter from emptyLogRecordArguments

evanlauer1 committed:
Added LogProcessors and implemented shutdown and forceFlush for LoggerProviders (#132)

* Renamed LogProcessor to LogRecordProcessor

* Added `noOpLoggerProvider` and removed `MonadIO` constraint from `createLoggerProvider`

* Renamed `LogProcessor` to `LogRecordProcessor`

* Call processors' onEmit when a log is emitted

evanlauer1 committed:
Renamed Processor to SpanProcessor, Exporter to SpanExporter, and Logging to Logs (#134)

* Renamed Processor to SpanProcessor, Exporter to SpanExporter, and Logging to Logs

* Updated .cabal files

* Updated hie.yaml

* Updated nix haskell packages

* Renamed span-exporter modules to exporter

* Continued module restructuring

* Finished module restructuring

* Renamed SpanProcessor to Processor.Span

* Removed LogRecord from Legacy Processor and Exporter modules

* Added pattern Synonyms for Exporters and Processors

* Added deprecated warnings to Exporters and Processors

* Fixed nix haskell-packages

* Fixed typo in Zipkin exporter

Commit:
add protobuf nix dependency to stack.yaml (#140)

evanlauer1 committed:
Added LogRecordExporters and updated documentation. (#135)

* Added LogRecordExporters and updated documentation.

* Fixed CI not passing

* Removed HashMap from logRecordExporterExport

* Wrap LogRecordExporter in MVar so it cannot be called concurrently

* Added documentation

* Renamed LogRecordExporterInternal to LogRecordExporterArguments and updated documentation

Commit:
add C libraries to flake.nix (#143)

* add C libraries to flake.nix

* add grpc to shell

magthe committed:
Instrument postgresql simple (#11)

* [psql-simple] Add a custom span function to ease instrumention

* [psql-simple] Instrument Database.PostgreSQL.Simple.query*

* [psql-simple] Instrument Database.PostgreSQL.Simple.execute*

* [psql-simple] Instrument Database.PostgreSQL.Simple.fold*

* [psql-simple] Instrument Database.PostgreSQL.Simple.forEach*

* [psql-simple] Instrument Database.PostgreSQL.Simple.returning*

* [psql-simple] Export the utility function pgsSpan

* [psql-simple] Add HasCallStack on all functions

velveteer committed:
Bracket WAI middleware spans with detachContext (#116)

After updating to hs-opentelemetry-instrumentation-wai-0.1.0.0, I
noticed our Datadog APM traces were starting to group unrelated requests
together. Notably all of these sporadic groups shared the same GHC
thread ID.

Further investigation led me to this commit: 34569fc. At first glance,
removing this bracket would seem harmless, assuming each WAI request is
served by a separate thread. But in fact warp uses some heuristics to
enable the HTTP keep-alive optimization, which means a single thread can
serve multiple requests. Without detaching the context from the thread
after serving a request, the initial request context is re-used for
subsequent requests, which means propagated OTel context will be
inherited by those request spans.

I tested this small patch in our production application and it appears
to have resolved our APM woes.

Also see GitHub issue #130.

pbrisbin committed:
Set exception details when using WAI and Yesod instrumentation together (#121)

* Address warnings and HLint

* Extract isInternalError

* Set exception details when using WAI and Yesod instrumentation together

When the WAI instrumentation is used (so there is a span on the
request), this instrumentation only sets the shared attributes. Since
exceptions never reach WAI when used with Yesod (since they're caught to
error pages), this means that error traces don't include any details. By
calling `recordException` and `setStatus` here, we can address that.

Fixes #102

* Fix style

* Add comment about explicit exception marking

marinelli committed:
Bump OTLP specification version to v1.0.0 (#150)

* Bump OTLP

* Fix building

* Add generate-protocol.sh script

* Update generate-protocol.sh script

* Fix version

* Improve generate-protocol.sh script

* Improve generate-protocol.sh script

* Improve generate-protocol.sh script

* Switch to OTLP v1.0.0

* Remove generate-protocol.sh script

* Update flake.lock

* Update otlp/package.yaml

* Update ChangeLog.md and bump package version

* Fix hs-opentelemetry-exporter-otlp constraints

* Revert packages versions bumps and update otlp/ChangeLog.md

* Revert hs-opentelemetry-otlp version bump

develop7 committed:
Enable it build on Windows by `unix` → `unix-compat` transition (#114)

* Introduces compilation on Windows support

Extracts platform-dependent code to OpenTelemetry.Platform module

* sdk: update ChangeLog.md

marinelli committed:
Fix reading boolean environment variables (#153)

* Fix reading boolean environment variables

* Apply review suggestions

* Apply review suggestions

* Follow specifications for environment boolean values

michaelpj committed:
Support OTEL_SDK_DISABLED (#148)

Very simply, we return the empty tracer provider options in this case,
which makes tracing do nothing.

michaelpj committed:
Fix attribute length limits (#151)

Fixes #113

michaelpj committed:
Add the ability to add links to spans after creation (#152)

ehofreiter committed:
Fix resource merge (#156)

Fixes #111

michaelpj committed:
Add instrumentation for tasty (#149)

This adds spans for all of:
1. Individual test cases
2. Test groups
3. Resource setup and teardown

It also correctly handles test parallelism.

marinelli committed:
UPKEEP: Clean up dependencies (#155)

* Lint packages dependencies; remove unused imports

* Fix package dependencies after rebasing

* Remove commented imports

* Revert import of package-common.yaml for hs-opentelemetry-exporter-handle

* Minor linting

* Revert useless changes

* Use underscore to disable not implemented test suites or libraries

marinelli committed:
Use a shell script for generating OTLP modules (#159)

* Add generate-protocol.sh script

* Simplify and rename script

* Remove custom setup

* Generate Haskell modules with the script

* Remove Setup.hs file

* Add .gitattributes

* Use OTLP_VERSION file

* Update README.md

* Update flake.nix

* Update .gitattributes

michaelpj committed:
Add a TH helper for creating InstrumentationLibrarys (#161)

michaelpj committed:
Avoid dependency on the SDK for tasty instrumentation (#165)

This is unnecessary, and instrumentation shouldn't depend on the SDK.

marinelli committed:
Delete placeholder packages (#166)

* Remove hs-opentelemetry-propagator-jaeger

* Remove hs-opentelemetry-exporter-zipkin

* Remove hs-opentelemetry-exporter-jaeger

* Remove hs-opentelemetry-exporter-prometheus

michaelpj committed:
Prepare release (#164)

* Update changelogs

* Bump API and SDK versions

* Bump OTLP version

* Update versions for exporters and instrumentation

* Changelog improvements

* Update upload script

michaelpj committed:
Release fixes (#168)

Commit:
remove a duplicated ChangeLog file (#169)

Commit:
[api] attribute map and type-safe attribute (#157)

* [api] attribute map and type-safe attribute

* fix haddock comment

Co-authored-by: ehofreiter <35861161+ehofreiter@users.noreply.github.com>

* fix haddock comment

Co-authored-by: ehofreiter <35861161+ehofreiter@users.noreply.github.com>

* add comment

Co-authored-by: Michael Peyton Jones <me@michaelpj.com>

* rename Key to AttributeKey

* make better description

---------

Co-authored-by: ehofreiter <35861161+ehofreiter@users.noreply.github.com>
Co-authored-by: Michael Peyton Jones <me@michaelpj.com>

Commit:
fix ignoreGeneratedFiles function (#170)

danidiaz committed:
Make propagators use RequestHeaders as outbound carriers. (#171)

Fixes #160.

Co-authored-by: Daniel Díaz <diaz_carrete@yahoo.com>

Commit:
[semantic-conventions] init (#142)

michaelpj committed:
Improve source code attribute handling (#172)

* Improve source code attribute handling

My previous attempt at this was wrong for a few reasons.
1. I got the locations misaligned within the call stack, so we were
   giving the wrong call location.
2. It's in fact just totally unclear what to do with frozen call stacks
   in general.

This commit fixes 1 and enables a different approach to 2: get the user
to explicitly supply source code attributes if they are doing something
clever. So rather than trying to freeze the call stack to get us to do
waht you want, you can just create the source code attributes yourself
(using `callerAttributes` if you like) and add them. The main change to
enable this is just to not add source code attributes from the call
stack if there are any already present, which prevents us getting a
muddled mix of attributs if only some have been set.

* Add an example test with a span creation helper

turnage committed:
Export `toImmutableSpan` (#174)

iand675 committed:
Add sqlcommenter to cabal.project

iand675 commented on utils/sqlcommenter/src/SqlCommenter.hs:
```diff
@@ -0,0 +1,268 @@
+
+intercalate :: (Monoid a, Foldable f) => a -> f a -> a
+intercalate delim l = mconcat (intersperse delim l)
+{-# INLINE intercalate #-}
```
These are foldable versions of the list equivalent.

iand675 commented on utils/sqlcommenter/src/SqlCommenter.hs:
```diff
@@ -0,0 +1,268 @@
+-- SQL Commenter spec wants them escaped with a slash, but this should
+-- probably solve the same issue
+unreservedQS :: Word8Set
+unreservedQS = foldr insert SqlCommenter.empty $ map c2w "-_.~'"
```
This builds a data structure that can perform fast checks to determine whether to escape a character or not for a URL. This is similar to @ekmett's charset package, but with a much more constrained purposed, and is intended to avoid incurring extra dependencies. https://hackage.haskell.org/package/charset-0.3.11/docs/Data-CharSet.html

jkachmar commented on utils/sqlcommenter/src/SqlCommenter.hs:
```diff
@@ -0,0 +1,268 @@
+import OpenTelemetry.Trace.Core as Core
+import System.IO.Unsafe
+
+sqlCommenterKey :: Ctxt.Key (Map Text Text)
+sqlCommenterKey = unsafePerformIO $ Ctxt.newKey "sqlcommenter-attributes"
+{-# NOINLINE sqlCommenterKey #-}
+
+lookupSqlCommenterAttributes :: Ctxt.Context -> Map Text Text
+lookupSqlCommenterAttributes ctxt = case Ctxt.lookup sqlCommenterKey ctxt of
+  Nothing -> mempty
+  Just attrs -> attrs
+
+getSqlCommenterAttributes :: IO (Map Text Text)
+getSqlCommenterAttributes = lookupSqlCommenterAttributes <$> TL.getContext
```
I guess `sqlCommenterKey` has to be a global variable b/c of how `Vault` works, and since it's an implementation detail that never leaks beyond the scope of this module there shouldn't be a problem here?

checking my understanding: we can't make `sqlCommenterKey :: IO _` because we need `lookupSqlCommenterAttributes` to have a pure API for any callers who might need to use it in a pure context (even though `getSqlCommenterAttributes :: IO _`).

iand675 replied:
Right. The main thing is that `sqlCommenterKey` needs to be _stable_ across uses. If we had users intitialize the key and pass it in any time they want to use it to look up the sqlcommenter attributes, then we could make it be an IO value. However, I don't think that's a great experience, and is relatively error prone compared to just making sure that there's a stable reference available.

jkachmar replied:
I didn't see this sort of context in `newKey` docs, so it might be worth adding to a Haddock on `sqlCommenterKey` at least. up to you, though

jkachmar commented on utils/sqlcommenter/src/SqlCommenter.hs:
```diff
@@ -0,0 +1,268 @@
+    & M.insert "traceparent" (unsafeConvert traceparent)
+    & M.insert "tracestate" (unsafeConvert tracestate)
+  where
+    unsafeConvert = B.toText . B.unsafeFromByteString
```
[encodeSpanContext](https://hackage.haskell.org/package/hs-opentelemetry-propagator-w3c-0.0.1.4/docs/OpenTelemetry-Propagator-W3CTraceContext.html#v:encodeSpanContext) gives us two `ByteString`s.

it looks like these are produced by [traceparentHeader](https://hackage.haskell.org/package/hs-opentelemetry-propagator-w3c-0.0.1.4/docs/src/OpenTelemetry.Propagator.W3CTraceContext.html#local-6989586621679050537), which appears to construct these directly using a combination of primitives (`char7 '-'`, `word8HexFixed` on the `Word8` stored in `TraceFlags`) & `Base16`-encoded `SpanId` & `TraceId` values.

You've got tests for this, so presumably something upholds the invariant that all of these must be UTF-8 encoded text but I'm having trouble understand what exactly this is (if it is anything besides convention).

iand675 replied:
Good question. The W3C `traceparent` and `tracestate` headers [by definition](https://www.w3.org/TR/trace-context/#traceparent-header-field-values) are US-ASCII encoded, so we don't have to think super hard about this assuming that `import OpenTelemetry.Propagator.W3CTraceContext (encodeSpanContext)` is implemented correctly.

jkachmar replied:
aha okay, that's the missing link for me; I would maybe add this in a comment onto `unsafeConvert` in the same manner that Rust encourages documenting assumed invariants around `unsafe` blocks just so that it's clear.

jkachmar commented on utils/sqlcommenter/test/Spec.hs:
```diff
@@ -0,0 +1,52 @@
+
+main :: IO ()
+main = hspec $ do
+  specify "Does not add comment if the query already has a comment" $ do
+    let queries =
+          [ "SELECT * FROM table -- noodle"
+          , "SELECT * FROM table -- noodle\n"
+          , "SELECT * FROM table /* noodle\n  poodle\n*/"
+          ]
+        someAttrs = M.fromList [("foo", "bar")]
+    for_ queries $ \query -> do
+      sqlCommenter query someAttrs `shouldBe` query
```
I don't quite follow what this is testing; `someAttrs` is just ignored entirely and the query is returned as-is because there is an existing comment?

iand675 replied:
Yes, exactly.

Per the sqlcommenter spec:

> If a comment already exists within a SQL statement, we MUST NOT mutate that statement.

jkachmar commented on utils/sqlcommenter/test/Spec.hs:
```diff
@@ -0,0 +1,52 @@
+    hedgehog $ do
+      t <- forAll textValGen
+      t === (T.decodeUtf8 $ urlDecode True $ T.encodeUtf8 $ SqlCommenter.urlEncode t)
+  specify "Span attributes are picked up from thread-local context" $ do
+    tp <- createTracerProvider [] emptyTracerProviderOptions
+    let t = makeTracer tp (InstrumentationLibrary "test" "test") tracerOptions
+    inSpan t "test" defaultSpanArguments $ do
+      attrs <- getSqlCommenterAttributesWithTraceData
+      M.lookup "traceparent" attrs `shouldSatisfy` isJust
+      M.lookup "tracestate" attrs `shouldSatisfy` isJust
+  specify "Parsing a queries reads attributes from the first comment" $ do
+    let query1 = "SELECT * FROM table -- foo='bar'\n-- bar='baz'"
+        query2 = "SELECT * FROM table /* noodle='poodle',wibble='wobble'*/ /* foo='bar' */"
+    parseFirstSqlComment query1 `shouldBe` M.fromList [("foo", "bar")]
+    parseFirstSqlComment query2 `shouldBe` M.fromList [("noodle", "poodle"), ("wibble", "wobble")]
+  specify "Parsing a query decodes encoded bytes" $ do
+    hedgehog $ do
+      let attrList = M.fromList <$> Gen.list (Range.linear 0 30) ((,) <$> textValGen <*> textValGen)
+      kvs <- forAll attrList
+      kvs === parseFirstSqlComment (sqlCommenter "" kvs)
```
I don't _think_ any of these tests cover the case where `sqlCommenter` actually modifies the query, right?

I think it'd be useful to add an end-to-end to check that span attributes get picked up tacked onto a query in comments according to the [sqlcommenter example documentation](https://google.github.io/sqlcommenter/) (or something).

iand675 replied:
Very good point. Will add tests for this.

michaelpj commented:
I think this could do with a bit of explanation for how to actually _use_ it? As far as I can tell, there are several things here:
1. Support for adding `sqlcommenter` comments to queries. This seems useful and arguably deserves its own library, which need not depend on any OTel bits?
2. Support for a named section of the thread-local context which is used for special attributes associated with `sqlcommenter`
3. Support for adding some OTel specific attributes into the special section

I guess the expectation is that on top of this you:
1. Modify the instrumentation libraries for databases to call `getSqlCommenterAttributesWithTraceData` and then use `sqlcommenter` to actually put them in queries.
2. Tell people to add things into the special `sqlcommenter` section in order to get them into their queries.

I guess I'm a bit unsure about the "action-at-a-distance" transmission of the attributes. This requires knowing to put attributes in a special place, that is a) specific to databases, and b) specific to a particular method of adding information into queries. As a user, I maybe don't even know that the `postgresql-simple` instrumentation library is using `sqlcommenter` - should I need to know that? If we were going to have a special section for such attributes, perhaps it should be called `postgres` or `database`?

An alternative would be to use something like the existing support for ["carry-on" attributes](https://hackage.haskell.org/package/hs-opentelemetry-api-0.2.0.0/docs/OpenTelemetry-Contrib-CarryOns.html). Then if a higher-scope piece of code wants to add an attribute that should get into the attributes of a database span, it can add it to the carry-ons?

jkachmar committed:
drop `http-client` MonadUnliftIO constraint (#177)

* chore: delete `grep` file

* feat(http-client): drop 'MonadUnliftIO' constraint

the `http-client` wrapper functions should all work just fine in plain
`IO`, so we should be able to present an API that is more aligned with
the upstream functions they are wrapping.

jkachmar committed:
revert "drop `http-client` MonadUnliftIO constraint (#177)" (#178)

This partially reverts e514419 to avoid breaking changes to the API.

marinelli committed:
Fix building with the latest text and persistent releases (#179)

iand675 committed:
Fix tests

iand675 committed:
Add some sqlcommenter docs, and export some more functions

michaelpj commented:
We discussed this at Zurihac, let me try and summarise what we concluded:
- The `sqlcommenter` behaviour looks a lot like a kind of `Propagator`
- At the moment, our ability to globally configure propagators is a bit limited: we can only configure ones that propagate into HTTP headers, as opposed to, say, queries.
- If we could globally configure "query propagators" then we could have a nice workflow:
    - Database instrumentation libraries could look up any appropriate configured query propagators and apply them
         - What does this look like? do we need per-database-type propagators? probably? Or could they just be key-value propagators that get then applied with `sqlcommenter`?
    - If you want to use `sqlcommenter`, you can add a query propagator that looks up the special attributes in the context and puts them in. So the spookiness is isolated to the definition of this propagator, which is optional.
    - If you want to do something else with `sqlcommenter`, you can write your own propagator that does something different.
- One way to configure the propagators in a more forwards-compatible way would be to have a `vault` of propagators, so different libraries can have well-known keys that let you look up specific propagators if they're present.

I am wondering whether it would indeed be nice to consider the query propagators as just key-value map propagators. Then `sqlcommenter` is just the implementation of how those maps then get put into the actual query. I think I need to take another look at what the spec says!

jkachmar commented:
superseded by #311
